File: eclipse-plugin/src-ui/jche/eclipse/server/Wire.java
// Copyright 2026 Inoue Kazuhiro (instreest). SPDX-License-Identifier: Apache-2.0
package jche.eclipse.server;

/**
 * プロトコルの文字列の逃がし方。サーバー側の {@code jche.server.Protocol} と対になる。
 *
 * <p>同じ規則を2か所に書いているのは、プラグイン側が解析本体（{@code jche-core.jar}）を
 * <b>クラスパスに載せない</b>ためである（載せると Eclipse の JDK で読まれてしまい、
 * Java 8 で動かすという目的が崩れる）。規則は短いので、写す方が依存を作るより安い。
 * 変えるときは両方を直すこと（{@code test/plugin-client/run.sh} が往復で検査する）。
 */
final class Wire {

    static final String SEP = "\t";

    private Wire() {
    }

    static String escape(String value) {
        if (value == null || value.isEmpty()) {
            return "";
        }
        StringBuilder sb = new StringBuilder(value.length() + 8);
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            if (c == '\\') {
                sb.append("\\\\");
            } else if (c == '\t') {
                sb.append("\\t");
            } else if (c == '\n') {
                sb.append("\\n");
            } else if (c == '\r') {
                sb.append("\\r");
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    static String unescape(String value) {
        if (value == null) {
            return "";
        }
        if (value.indexOf('\\') < 0) {
            return value;
        }
        StringBuilder sb = new StringBuilder(value.length());
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            if (c != '\\' || i + 1 >= value.length()) {
                sb.append(c);
                continue;
            }
            char next = value.charAt(++i);
            if (next == 't') {
                sb.append('\t');
            } else if (next == 'n') {
                sb.append('\n');
            } else if (next == 'r') {
                sb.append('\r');
            } else if (next == '\\') {
                sb.append('\\');
            } else {
                sb.append(next);
            }
        }
        return sb.toString();
    }
}

File: vscode-plugin/src/server/wire.ts
// Copyright 2026 Inoue Kazuhiro (instreest). SPDX-License-Identifier: Apache-2.0

/**
 * プロトコルの文字列の逃がし方。サーバー側の `jche.server.Protocol` と対になる。
 *
 * 同じ規則を Java（サーバー・Eclipse プラグイン）と TypeScript（ここ）に書いているのは、
 * 拡張ホストが Node.js で、解析本体のクラスを読めないからである。規則は短いので、
 * 写す方が依存を作るより安い。変えるときは3か所を直すこと
 * （`test/server/run.sh`・`test/plugin-client/run.sh`・`test/vscode/run.sh` が往復で検査する）。
 */
export const SEP = '\t';

/** TAB 区切りを壊す文字を逃がす。undefined は空文字にする */
export function escape(value: string | undefined): string {
    if (!value) {
        return '';
    }
    let out = '';
    for (const c of value) {
        switch (c) {
            case '\\': out += '\\\\'; break;
            case '\t': out += '\\t'; break;
            case '\n': out += '\\n'; break;
            case '\r': out += '\\r'; break;
            default: out += c;
        }
    }
    return out;
}

/** `escape` の逆。受け取り側（サーバー）と同じ規則であることが大事 */
export function unescape(value: string | undefined): string {
    if (!value) {
        return '';
    }
    if (value.indexOf('\\') < 0) {
        return value;
    }
    let out = '';
    for (let i = 0; i < value.length; i++) {
        const c = value[i];
        if (c !== '\\' || i + 1 >= value.length) {
            out += c;
            continue;
        }
        const next = value[++i];
        switch (next) {
            case 't': out += '\t'; break;
            case 'n': out += '\n'; break;
            case 'r': out += '\r'; break;
            case '\\': out += '\\'; break;
            default: out += next;
        }
    }
    return out;
}

/** 要求の語を TAB でつなぐ。語の中の TAB・改行は逃がす */
export function join(words: readonly string[]): string {
    return words.map(escape).join(SEP);
}
